File: python/src/day10/main.py
from src.common.file_utils import get_path, read_integers


def part_one(filename: str) -> int:
    ints = read_integers(get_path(__file__, filename))
    ints.sort()
    ones, threes, v = 0, 0, 0

    for i, x in enumerate(ints):
        if x - v == 3:
            threes += 1
        if x - v == 1:
            ones += 1
        v = x
    
    return ones * (threes + 1)


def part_two(filename: str) -> int:
    ints = read_integers(get_path(__file__, filename))
    ints.sort()
    max_ints = max(ints)
    cache = [-1] * max_ints

    def number_ways(v):
        if v not in ints:
            return 0
        if v == max_ints:
            return 1

        if cache[v] == -1:
            s = sum([number_ways(v + x) for x in range(1, 4)])
            cache[v] = s

        return cache[v]

    return sum([number_ways(x) for x in range(1, 4)])


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)

File: go/day10/main.go
package main

import (
	input "aoc2020/inpututils"

	"fmt"
	"sort"
)

func main() {
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))

	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

func Part1(filename string) int {
	nums := input.ReadNumbers(filename)
	sort.Ints(nums)

	ones, threes, v := 0, 0, 0

	for _, x := range nums {
		if (x - v) == 1 {
			ones += 1
		} else if (x - v) == 3 {
			threes += 1
		}
		v = x
	}
	return ones * (threes + 1)
}

func Part2(filename string) int {
	nums := input.ReadNumbers(filename)

	numSet := map[int]struct{}{}
	for _, n := range nums {
		numSet[n] = struct{}{}
	}

	maxV := max(nums)
	cache := make([]int, maxV)
	for i := range cache {
		cache[i] = -1
	}

	var numberOfWays func(int) int

	numberOfWays = func(v int) int {
		if _, ok := numSet[v]; !ok {
			return 0
		}
		if v == maxV {
			return 1
		}
		if cache[v] == -1 {
			s := 0
			for _, x := range []int{1, 2, 3} {
				s += numberOfWays(v + x)
			}
			cache[v] = s
		}
		return cache[v]
	}
	return numberOfWays(1) + numberOfWays(2) + numberOfWays(3)
}

func max(s []int) int {
	m := s[0]
	for _, v := range s {
		if v > m {
			m = v
		}
	}
	return m
}
